File: MarxUp/MarxUp/DataSources/ToolboxItemCollectionDataSource/ShapesCollectionViewDataSource.m
//
//  ShapesCollectionViewDataSource.m
//  MarxUp
//
//  Created by Ognyanka Boneva on 26.09.18.
//  Copyright © 2018 Ognyanka Boneva. All rights reserved.
//

#import "ShapesCollectionViewDataSource.h"
#import "ToolboxItemOptionsCollectionViewCell.h"
#import "ShapeModel.h"

@interface ShapesCollectionViewDataSource ()

@property (strong, nonatomic) NSArray<ShapeModel *> *shapes;

@end

@implementation ShapesCollectionViewDataSource

+ (instancetype)newDataSource {
    ShapesCollectionViewDataSource *newSource = [ShapesCollectionViewDataSource new];
    if (newSource) {
        ShapeModel *circle = [ShapeModel newShapeFromType:ShapeTypeCircle];
        ShapeModel *triangle = [ShapeModel newShapeFromType:ShapeTypeTriangle];
        ShapeModel *rectangle = [ShapeModel newShapeFromType:ShapeTypeRegularRectangle];
        ShapeModel *roundedRect  =[ShapeModel newShapeFromType:ShapeTypeRoundedRectangle];
        newSource.shapes = @[roundedRect, circle, triangle, rectangle];
    }
    
    return newSource;
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.shapes.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    ToolboxItemOptionsCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:IDENTIFIER_TOOLBOX_ITEM_OPTIONS_CELL forIndexPath:indexPath];
    ShapeModel *model = self.shapes[indexPath.item];
    
    cell.imageView.image = model.image;
    
    return cell;
}

- (ShapeType)optionAtIndex:(NSInteger)index {
    return self.shapes[index].type;
}


@end

File: MarxUpUp/MarxUpUp/DataSources/ToolboxItemCollectionDataSource/ShapesCollectionViewDataSource.swift
//
//  ShapesCollectionViewDataSource.swift
//  MarxUpUp
//
//  Created by Ognyanka Boneva on 14.11.18.
//  Copyright © 2018 Ognyanka Boneva. All rights reserved.
//

import UIKit

final class ShapesCollectionViewDataSource: NSObject {
    private let shapes = [ShapeModel(withShapeType: ShapeType.Circle), ShapeModel(withShapeType: ShapeType.RoundedRectangle), ShapeModel(withShapeType: ShapeType.RegularRectangle)]
}

extension ShapesCollectionViewDataSource : ToolboxItemCollectionDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return shapes.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(fromClass: ToolboxItemOptionsCollectionViewCell.self, forIndexPath: indexPath)
        
        cell.itemImageView.image = shapes[indexPath.item].image
        
        return cell
    }
    
    func option(atIndex index: Int) -> Int {
        return shapes[index].type.rawValue
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
}
